src/sdk/pynni/nni/compression/torch/pruning/admm_pruner.py

```diff
@@ -0,0 +1,197 @@
+        self._optimize_iterations = optimize_iterations
+        self._training_epochs = training_epochs
+        self._row = row
+
+        self.set_wrappers_attribute("if_calculated", False)
+
+    def validate_config(self, model, config_list):
+        """
+        Parameters
+        ----------
+        model : torch.nn.module
+            Model to be pruned
+        config_list : list
+            List on pruning configs
+        """
+
+        if self._base_algo == 'level':
+            schema = CompressorSchema([{
+                'sparsity': And(float, lambda n: 0 < n < 1),
+                Optional('op_types'): [str],
+                Optional('op_names'): [str],
+            }], model, _logger)
+        elif self._base_algo in ['l1', 'l2']:
+            schema = CompressorSchema([{
+                'sparsity': And(float, lambda n: 0 < n < 1),
+                'op_types': ['Conv2d'],
+                Optional('op_names'): [str]
+            }], model, _logger)
+
+        schema.validate(config_list)
+
+    def _projection(self, weight, sparsity):
+        '''
+        Return the Euclidean projection of the weight matrix according to the pruning mode.
+
+        Parameters
+        ----------
+        weight : tensor
+            original matrix
+        sparsity : float
+            the ratio of parameters which need to be set to zero
+
+        Returns
+        -------
+        tensor
+            the projected matrix
+        '''
+        w_abs = weight.abs()
+        if self._base_algo == 'level':
+            k = int(weight.numel() * sparsity)
+            if k == 0:
+                mask_weight = torch.ones(weight.shape).type_as(weight)
+            else:
+                threshold = torch.topk(w_abs.view(-1), k, largest=False)[0].max()
+                mask_weight = torch.gt(w_abs, threshold).type_as(weight)
+        elif self._base_algo in ['l1', 'l2']:
+            filters = weight.size(0)
+            num_prune = int(filters * sparsity)
+            if filters < 2 or num_prune < 1:
+                mask_weight = torch.ones(weight.size()).type_as(weight).detach()
+            else:
+                w_abs_structured = w_abs.view(filters, -1).sum(dim=1)
+                threshold = torch.topk(w_abs_structured.view(-1), num_prune, largest=False)[0].max()
+                mask_weight = torch.gt(w_abs_structured, threshold)[:, None, None, None].expand_as(weight).type_as(weight)
+
+        return weight.data.mul(mask_weight)
+
+    def calc_mask(self, wrapper, **kwargs):
+        """
+        Calculate the mask of given layer.
+        Use the function of LevelPruner or L1FilterPruner according to the pruning mode.
+
+        Parameters
+        ----------
+        wrapper : Module
+            the module to instrument the compression operation
+        Returns
+        -------
+        dict
+            dictionary for storing masks
+        """
+        self.masker = MASKER_DICT[self._base_algo](self.bound_model, self)
```

Comment:
This `self.masker` is not used.

Reply:
> This `self.masker` is not used.
`self.masker` is required by OneShotPruner to realize `calc_mask`

Reply:
Moved to `__init__` .